Source/Core/Duality/Components/Physics/RigidBody.cs

```diff
@@ -772,50 +775,65 @@ public bool PickShapes(Vector2 worldCoord, Vector2 size, List<ShapeInfo> pickedS
 		{
 			if (this.body == null) return false;
 
-			Vector2 fsWorldCoord = PhysicsUnit.LengthToPhysical * worldCoord;
-			FarseerPhysics.Collision.AABB fsWorldAABB = new FarseerPhysics.Collision.AABB(fsWorldCoord, PhysicsUnit.LengthToPhysical * (worldCoord + size));
-
+			PolygonShape boxShape = new PolygonShape(new Vertices(new List<Vector2>() {
+				PhysicsUnit.LengthToPhysical * worldCoord,
+				PhysicsUnit.LengthToPhysical * new Vector2(worldCoord.X + size.X, worldCoord.Y),
+				PhysicsUnit.LengthToPhysical * (worldCoord + size),
+				PhysicsUnit.LengthToPhysical * new Vector2(worldCoord.X, worldCoord.Y + size.Y) }), 1);
+			Mat22 rot = new Mat22();
+			rot.SetIdentity();
+			Vector2 pos = new Vector2(0, 0);
+			FarseerPhysics.Common.Transform boxTransform = new FarseerPhysics.Common.Transform(ref pos, ref rot);
+			Manifold manifold = new Manifold();
 			int oldCount = pickedShapes.Count;
+
+			FarseerPhysics.Common.Transform bodyTransform;
+			this.body.GetTransform(out bodyTransform);
+
 			for (int i = 0; i < this.body.FixtureList.Count; i++)
 			{
 				Fixture f = this.body.FixtureList[i];
 				ShapeInfo s = f.UserData as ShapeInfo;
-
-				FarseerPhysics.Collision.AABB fAABB;
-				FarseerPhysics.Common.Transform transform;
-				this.body.GetTransform(out transform);
-				f.Shape.ComputeAABB(out fAABB, ref transform, 0);
-				
-				if (fsWorldAABB.Contains(ref fAABB))
+				PolygonShape polygonShape = f.Shape as PolygonShape;
+				if (polygonShape != null)
 				{
-					pickedShapes.Add(s);
-					continue;
+					Collision.CollidePolygons(ref manifold, boxShape, ref boxTransform, polygonShape, ref bodyTransform);
+					if (manifold.PointCount > 0)
+					{
+						pickedShapes.Add(s);
+						continue;
+					}
 				}
```

Comment:
If it's a polygon shape, but does not have any collisions, it will continue checking whether it could be a circle shape next. Consider changing this to an if-else structure, or moving the continue statement out of the point count check!

Reply:
Ye missed that, it should obviously not do the circle check after it already done the polygon one.

Too bad we dont have C# 7 as pattern matching would make this so much more readable.

Reply:
Found out there is a ShapeType enum property on the fixture so I can use that in a switch. Much more readable.